### PR TITLE
Work on migrating to conan v2

### LINKED ISSUE #1: Update conan packages
spdlog needs to be updated

### LINKED ISSUE #6: compilation error on Ubuntu22
When I try to compile the software (master branch), I get:
```
ERROR: Package 'Qt-Color-Widgets/9f4e052@ess-dmsc/stable' not resolved: Qt-Color-Widgets/9f4e052@ess-dmsc/stable: Cannot load recipe.
Error loading conanfile at '/home/user/.conan2/p/Qt-Co32bc14517ff1e/e/conanfile.py': Unable to load conanfile in /home/user/.conan2/p/Qt-Co32bc14517ff1e/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/user/.conan2/p/Qt-Co32bc14517ff1e/e/conanfile.py", line 2, in <module>
    from conans import ConanFile, CMake, tools
ImportError: cannot import name 'ConanFile' from 'conans' (/usr/local/lib/python3.10/dist-packages/conans/__init__.py)
.
CMake Error at build/conan.cmake:540 (message):
  Conan install failed='1'
Call Stack (most recent call first):
  build/conan.cmake:882 (old_conan_cmake_install)
  CMakeLists.txt:32 (conan_cmake_run)
```

### PR DESCRIPTION
**Pull request by php1ic**
Updates on the road to getting things working with conan v2.

This should **NOT** yet be merged as it doesn't compile. There are include errors relating to 1 of the packages pulled in by conan (nlohmann) but I can't see why that package is any different to the others so need some help.

### CODE REVIEW

File: README.md
Comment (ferdymercury) [inline README.md]: I think it would be great to specify a oneliner command doing the git clone of already the specific branch.

**Comment by ferdymercury** [comment]
Maybe this helps?
https://github.com/nlohmann/json/discussions/3483
https://json.nlohmann.me/integration/package_managers/#conan

**Comment by php1ic** [comment]
The branch now compiles and I can run the program :+1: 

In the top level CMakeLists file I had to explicitly call `find_package()` on all the packages installed via conan then make sure they were linked against when building the main executable. Perhaps in conan v1 a lot of that was automagically done by the code used in cmake and the file it downloaded?

File: README.md
Comment (php1ic) [reply README.md]: How about adding the cmake-conan repo as a submodule? We can then set the cmake option in the CMakeLists file as we know the full path to the necessary file.

Comment (ferdymercury) [reply README.md]: Sounds good! We could also try FetchContent instead of a submodule? https://cmake.org/pipermail/cmake/2019-June/069709.html

Comment (martukas) [reply README.md]: I would to not have either submodule or assumption of anything cloned or installed to the system that does not come through a common package manager such as apt/pip/brew. From my perspective `cmake-conan` loses its utility for conan2, the way they have made it.
How about just having the user run the conan command manually prior to building? That's one extra line, but that's still less than submodule shenanigans.

File: .gitignore
Comment (martukas) [inline .gitignore:21]: what's lsp?

**Review by martukas — CHANGES_REQUESTED** [review]
I would update the CI script as well to build it in the new way. It's a good way to make sure it works in general, not just on your machine. It's also blocking the merge.

File: .gitignore
Comment (ferdymercury) [reply .gitignore:21]: https://github.com/emacs-lsp/lsp-mode/discussions/3478 maybe?

File: README.md
Comment (ferdymercury) [reply README.md]: Thanks php1ic.  I tried your branch, however I cannot compile on Ubuntu 22, it says:

```
======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to /tmp/nuclei/build/build/conan
conanfile.txt: Generator 'CMakeDeps' calling 'generate()'
conanfile.txt: Generator 'CMakeToolchain' calling 'generate()'
conanfile.txt: CMakeToolchain generated: conan_toolchain.cmake
conanfile.txt: Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.txt: If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/tmp/nuclei/build/build/conan/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.txt: CMakeToolchain generated: CMakePresets.json
conanfile.txt: CMakeToolchain generated: ../../../CMakeUserPresets.json
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
Install finished successfully
-- CMake-Conan: CONAN_GENERATORS_FOLDER=/tmp/nuclei/build/build/conan
-- CMake-Conan: CONANFILE=/tmp/nuclei/conanfile.txt
CMake Error at /tmp/cmake-conan/conan_provider.cmake:278 (find_package):
  Could not find a package configuration file provided by "Threads" with any
  of the following names:

    ThreadsConfig.cmake
    threads-config.cmake

  Add the installation prefix of "Threads" to CMAKE_PREFIX_PATH or set
  "Threads_DIR" to a directory containing one of the above files.  If
  "Threads" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:7 (find_package)

```

Comment (ferdymercury) [reply README.md]: > I would to not have either submodule or assumption of anything cloned or installed to the system that does not come through a common package manager such as apt/pip/brew. From my perspective `cmake-conan` loses its utility for conan2, the way they have made it. How about just having the user run the conan command manually prior to building? That's one extra line, but that's still less than submodule shenanigans.

Sounds good to me. What would be then the exact install command for conan?

File: .gitignore
Comment (php1ic) [reply .gitignore:21]: Yep, Language Server Protocol - https://en.wikipedia.org/wiki/Language_Server_Protocol & https://neovim.io/doc/user/lsp.html

In this case the file is used by [clangd](https://clangd.llvm.org/) so that may be a better name to use in the comment.

File: README.md
Comment (php1ic) [reply README.md]: This configuration issue with the `Threads` package could be another reason to move away from `cmake-conan`. When I made the PR the tip of the `develop2` branch was at [38a78aaf](https://github.com/conan-io/cmake-conan/commit/38a78aaf3a5567bd29932bf23e9) but the tip now also gives me that error.

Running conan externally, then cmake, does not give this issue. I'll update the README with the conan command to run and modified cmake commands.

CMake introduced [presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) to simply configuration and conan 2 makes use of that. It arrived in cmake 3.19, the newest release is 3.26 and the latest [ubuntu LTS](https://packages.ubuntu.com/search?keywords=cmake) is on 3.22. Any issues with me using presets in the instructions?

Comment (ferdymercury) [reply README.md]: Good idea the presets, thanks!

**Review by ferdymercury — APPROVED** [review]

File: README.md
Comment (php1ic) [reply README.md]: I should have read the conan output more carefully. Presets arrived in 3.19, but conan uses functionality from 3.23 so they can't be used by everyone at the moment.

The cmake command looks worse than it is, but if we want to make it super easy, we could write a simple build script?

File: CMakeLists.txt
Comment (martukas) [inline CMakeLists.txt]: Must it be 19 now? Might be too restrictive? Could it be just blanket 3.anything? Or do we do things that are only available in latest?

File: README.md
Comment (martukas) [inline README.md]: Wow, yeah this does look way more complicated than I'm used to. I understand "Release", but can't the other things just already be defined somewhere in the CMake files?
And this CMP0091 thing also looks like a sign of something funky. Does this have to do anything with accommodating old ABIs? I think the more common approach is to run `conan profile update settings.compiler.libcxx=libstdc++11 default`  after the conan config detect.

**Review by martukas — CHANGES_REQUESTED** [review]
Everything looks almost ready. And great to see CI building successfully now.
But the build process still actually looks more complicated than it should be. 
If that truly cannot be avoided then yes, a build scirpt is maybe a good idea. But I do suspect it can be a bit simplified?

File: README.md
Comment (martukas) [reply README.md]: Where is the dividing line for who has the new CMake? I don't see any need to actively support anything older than what the latest Ubuntu LTS has.

Comment (ferdymercury) [reply README.md]: Ubuntu LTS has 3.22 but conan uses 3.23, i think that was the issue

File: README.md
Comment (ferdymercury) [reply README.md]: I agree, I would suggest putting inside CMakeLists : if cmake version... Then set policy ..

And maybe also:
```
set(default_build_type "Release")
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
  message(STATUS "Setting build type to '${default_build_type}' as none was specified.")
  set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE STRING "Choose the type of build." FORCE)
  # Set the possible values of build type for cmake-gui
  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release")# "MinSizeRel" "RelWithDebInfo"
endif()
```

File: README.md
Comment (php1ic) [reply README.md]: Correct, 3.23 introduced functionality that conan makes use of and the Ubuntu LTS (and thus the CI image) is on 3.22

File: README.md
Comment (php1ic) [reply README.md]: >  I think the more common approach is to run `conan profile update settings.compiler.libcxx=libstdc++11 default` after the conan config detect.

This functionality was [removed](https://github.com/conan-io/conan/issues/12833) in v2

I copied the advice/instruction given by conan once it was run, but I think there like to be explicit, so we can safely drop 2 of the command line options as:
- The build type is checked and set in [EnsureBuildType.cmake](https://github.com/martukas/nuclei/blob/main/cmake/EnsureBuildType.cmake) so a new user doesn't need to explicitly set it.
- The `CMP0091` setting is [MSVC](https://cmake.org/cmake/help/latest/policy/CMP0091.html) specific. Looks like this project only checks for MSVC in the boost config and not in a way related to this setting.

The other option (`CMAKE_TOOLCHAIN_FILE`) tells cmake where the packages are installed so is definitely needs. We could assume the user has copy/pasted the command(s), so we know the paths and can set the option in the top level cmake file.

Another option could be to run conan from inside cmake with [execute_process](https://cmake.org/cmake/help/v3.22/command/execute_process.html), but this feels like we are re-inventing what [cmake-conan](https://github.com/conan-io/cmake-conan) is doing so there must be some intricacies I'm not aware of.

**Comment by php1ic** [comment]
Conan work has been moved into it's own cmake file and we are back to only needing to run `cmake ..` from the build directory.

Minor annoyance is that conan will always be called with cmake as I couldn't find a good way to check if packages are already installed thus stopping subsequent (unnecessary) calls to conan. If we don't want to do this, we can delete the new cmake file, remove the call from the top level cmake file and have a separate call to conan. The cmake call will still just be `cmake ..` as the variable is set externally to the conan specific file.

File: .gitignore
Comment (ferdymercury) [inline .gitignore]:
```suggestion
#Compilation database used by lsp (clangd)
```

File: CMakeLists.txt
Comment (ferdymercury) [inline CMakeLists.txt]:
```suggestion
cmake_minimum_required(VERSION 3.7.0 FATAL_ERROR)
```

**Comment by ferdymercury** [comment]
Builds well with Ubuntu 22, thanks for this!!

**Comment by php1ic** [comment]
I've made a PR on the [conan-Qt-Color-Widgets](https://github.com/ess-dmsc/conan-Qt-Color-Widgets/pull/8) repo to update to the latest commit on [Qt-Color-Widgets](https://github.com/ess-dmsc/Qt-Color-Widgets). Local testing of this repo with the latest commit doesn't raise any issues.

Should we wait for that PR to be merged then update the conan reference before accepting this merge? (assuming that PR is accepted of course :wink:  )

**Comment by ferdymercury** [comment]
@php1ic is this PR also addressing https://github.com/martukas/nuclei/issues/1 ?

**Comment by php1ic** [comment]
> @php1ic is this PR also addressing #1 ?

It wasn't the plan, but the build is now failing due to mismatches package dependencies, so I'll check things are the latest.

**Comment by php1ic** [comment]
The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.

I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

**Comment by ferdymercury** [comment]
> The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.

Thanks so much!!

> I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

When i run the CMake script on your current branch, it pulls 10.1.1 for fmt in my Ubuntu 22. So maybe spdlog depends on at least 10.1.1 and their docu is wrong?:

```
spdlog/1.12.0: Not found in local cache, looking in remotes...
spdlog/1.12.0: Checking remote: conancenter
spdlog/1.12.0: Downloaded recipe revision 0e390a2f5c3e96671d0857bc734e4731
fmt/10.1.1: Not found in local cache, looking in remotes...
fmt/10.1.1: Checking remote: conancenter
fmt/10.1.1: Downloaded recipe revision cd63809a79574a2f9eb73ca35f16a243
```

**Comment by ferdymercury** [comment]
> The spdlog package is now at the latest version and I have removed the explicit dependency on fmt as it is pulled in as part of spdlog. Everything else uses the latest versions.
> 
> I'm not sure why, but if I tried to specify `10.0.0` for fmt (i.e. the same version as the dependency for [spdlog/1.12.0](https://conan.io/center/recipes/spdlog?version=1.12.0)), there were conflict errors with `10.1.1`.

See https://github.com/conan-io/conan-center-index/commit/90e3ab49607d0f3ddaa9d5e53d87e367c45dbd69 so the webpage seems outdated?

**Comment by php1ic** [comment]
Good find! Looks like the webpage is wrong.

I'm happy to either put the explicit reference to fmt back in the `conanfile.txt` or leave it out as spdlog will pull it in.

Best practice is probably to have it explicitly, in case the project ever stops using spdlog?

**Comment by php1ic** [comment]
Hi @martukas, I believe all the changes you requested have been done. Can you please confirm/accept.

**Comment by martukas** [comment]
Thanks. Sorry for being so slow to respond. I will try to review this as soon as I can.

File: .circleci/config.yml
Comment (martukas) [inline .circleci/config.yml:15]: suggestion: we should add also publish these prereq installation lines to the main readme as an example of how to do it at least for Ubuntu, which is anyway the most likely linux system this would be build on

**Review by martukas — APPROVED** [review]
Looks great. Sorry I took so long to review. I also just successfully built and ran this on a Windows 11 machine with [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/).

One small suggestion is to add instructions for prereq installation to main README. Even to me it did no occur to look in the CI script 😆 

Good to go. Thanks a ton!

**Pull request merged**
